src/cargo/sources/git/oxide.rs

```diff
@@ -120,7 +109,103 @@ pub fn with_retry_and_progress(
+                    }
+                    git::protocol::handshake::Error::Transport(_) => {
+                        let msg = concat!("network failure seems to have happened\n",
+                            "if a proxy or similar is necessary `net.git-fetch-with-cli` may help here\n",
+                            "https://doc.rust-lang.org/cargo/reference/config.html#netgit-fetch-with-cli"
+                        );
+                        return Err(err.context(msg));
+                    }
+                    _ => None,
+                };
+                if let Some(auth_message) = auth_message {
+                    let mut msg = "failed to authenticate when downloading \
+                       repository"
+                        .to_string();
+                    if let Some(url) = last_url_for_authentication {
+                        msg.push_str(": ");
+                        msg.push_str(url.to_str_lossy().as_ref());
+                    }
+                    msg.push('\n');
+                    msg.push_str(auth_message);
+                    msg.push_str("\n\n");
+                    msg.push_str(
+                        "if the git CLI succeeds then `net.git-fetch-with-cli` may help here\n",
+                    );
+                    msg.push_str(
+                        "https://doc.rust-lang.org/cargo/reference/config.html#netgit-fetch-with-cli",
+                    );
+                    err = err.context(msg);
+                }
+            }
+            Err(err)
+        }
+    }
+}
+
+/// Produce a repository with everything pre-configured according to `config`. Most notably this includes
+/// transport configuration.
+pub fn open_repo(repo_path: &std::path::Path, config: &Config) -> CargoResult<git::Repository> {
+    git::open_opts(repo_path, {
+        let mut opts = git::open::Options::default();
+        let timeout = HttpTimeout::new(config)?;
+
+        // We need `git_binary` configuration as well for being able to see credential helpers
+        // that are configured with the `git` installation itself.
+        // However, this is slow on windows (~150ms) and most people won't need it as they use the
+        // standard index which won't ever need authentication.
+        // TODO: This is certainly something to make configurable, at the very least on windows.
+        //       Maybe it's also something that could be cached, all we need is the path to the configuration file
+        //       which usually doesn't change unless the installation changes. Maybe something keyed by the location of the
+        //       binary along with its fingerprint.
+        opts.permissions.config = git::permissions::Config::all();
+        opts.config_overrides([
+            format!("gitoxide.http.connectTimeout={}", timeout.dur.as_millis()),
+            format!("http.lowSpeedLimit={}", timeout.low_speed_limit),
+            format!("http.lowSpeedTime={}", timeout.dur.as_secs()),
+        ])
```

Comment:
And I agree, it feels fuzzy.

What let's me sleep at night is that these are considered part of the public API and I will do my best not to change them.

Longer term it would be possible and most definitely is desirable to [migrate the configuration documentation](https://github.com/Byron/gitoxide/blob/d336368ea2b6e5918b709220522cf1509ef30be2/src/plumbing/progress.rs#L93) into a data structure which can also be used for creating correctly named configuration. That way, users would have types and all the compile time checks going with it, along with documentation on `docs.rs`.

I added prototyping this on the TODO list, as it's a problem better to be solved (or improved) sooner than later.

Reply:
> What let's me sleep at night is that these are considered part of the public API and I will do my best not to change them.

If nothing else, they are at risk to typos, wrong types, etc.

I'm unsure if constants would help improve the situation or not.  If we feel they do, then a longer term step is making them type safe.  I can't remember how much I brought up about this but I dealt with an API with a lot of configuration and what we did was created constants that associated the primary key with the type in a way that special lookup functions would help type check things.  I want to play with that more for general git config work (granted, I'm using git2) but I wonder how well it could work here.

Reply:
By the sound of it, having compile-time checks will be an improvement so it's a go, even though it might not be the end of it in terms of usability.

I remember seeing the your work on putting git-configuration into the type system, but admittedly didn't spend the time to understand it. Lazy me immediately gravitated to "some documented strings are easy enough and do the job", ignoring it's disadvantages. That said, I think having the current configuration represented by types is a good tradeoff between complexity and safety to the point of being having no real disadvantages, and from there we can see what else is possible.

Reply:
For anyone following along, this commit has little changed but its resolved in a follow up commit